src/main/java/org/jenkinsci/plugins/docker/commons/impl/CompositeKeyMaterialFactory.java

```diff
@@ -10,70 +10,77 @@
 import java.io.Serializable;
 
 /**
- * Composes two {@link org.jenkinsci.plugins.docker.commons.KeyMaterialFactory}s into one.
+ * Composes multiple {@link org.jenkinsci.plugins.docker.commons.KeyMaterialFactory}s into one.
  *
  * @author Kohsuke Kawaguchi
  */
 @Restricted(NoExternalUse.class)
 public class CompositeKeyMaterialFactory extends KeyMaterialFactory {
-    private final KeyMaterialFactory lhs;
-    private final KeyMaterialFactory rhs;
+    private final KeyMaterialFactory[] factories;
 
-    public CompositeKeyMaterialFactory(KeyMaterialFactory lhs, KeyMaterialFactory rhs) {
-        this.lhs = lhs;
-        this.rhs = rhs;
+    public CompositeKeyMaterialFactory(KeyMaterialFactory... factories) {
+        this.factories = factories == null || factories.length == 0
+                ? new KeyMaterialFactory[]{new NullKeyMaterialFactory()}
+                : factories.clone();
     }
 
     @Override
     public KeyMaterial materialize() throws IOException, InterruptedException {
-        KeyMaterial lhsMaterialized = lhs.materialize();
-        KeyMaterial rhsMaterialized = null;
-        
+
+        KeyMaterial[] keyMaterials = new KeyMaterial[factories.length];
+        EnvVars env = new EnvVars();
         try {
-            rhsMaterialized = rhs.materialize();
-            EnvVars env = lhsMaterialized.env();
-            env.putAll(rhsMaterialized.env());
-    
-            return new CompositeKeyMaterial(env, lhsMaterialized, rhsMaterialized);
+            for (int index = 0; index < factories.length; index++) {
+                keyMaterials[index] = factories[index].materialize();
+                env.putAll(keyMaterials[index].env());
+            }
+            return new CompositeKeyMaterial(env, keyMaterials);
         } catch (Exception e) {
-            try {
-                lhsMaterialized.close();
-
-                // TODO Java 7+ use chained exceptions
-                if (e instanceof IOException) {
-                    throw (IOException) e;
-                } else if (e instanceof InterruptedException) {
-                    throw (InterruptedException) e;
-                } else {
-                    throw new IOException("Error materializing credentials.", e);
-                }
-            } finally {
-                if (rhsMaterialized != null) {
-                    rhsMaterialized.close();
+            for (int index = keyMaterials.length - 1; index >= 0; index--) {
+                try {
+                    if (keyMaterials[index] != null) {
+                        keyMaterials[index].close();
+                    }
+                } catch (IOException ioe) {
```

Comment:
In this case we are safely closing up any materialized instances because of a higher order problem. We want to throw the exception that was the root cause.